infrastructure/network/addressmanager/addressmanager.go

```diff
@@ -121,32 +146,61 @@ func (am *AddressManager) RemoveAddress(address *appmessage.NetAddress) error {
 	am.mutex.Lock()
 	defer am.mutex.Unlock()
 
+	return am.removeAddressNoLock(address)
+}
+
+// MarkConnectionFailure notifies the address manager that the given address
+// has failed to connect
+func (am *AddressManager) MarkConnectionFailure(address *appmessage.NetAddress) error {
+	am.mutex.Lock()
+	defer am.mutex.Unlock()
+
 	key := netAddressKey(address)
-	return am.store.remove(key)
+	entry, ok := am.store.getNotBanned(key)
+	if !ok {
+		return errors.Errorf("address %s is not registered with the address manager", address.TCPAddress())
+	}
+	entry.connectionFailedCount = entry.connectionFailedCount + 1
+	return am.store.updateNotBanned(key, entry)
+}
+
+// MarkConnectionSuccess notifies the address manager that the given address
+// has successfully connected
+func (am *AddressManager) MarkConnectionSuccess(address *appmessage.NetAddress) error {
+	am.mutex.Lock()
+	defer am.mutex.Unlock()
+
+	key := netAddressKey(address)
+	entry, ok := am.store.getNotBanned(key)
+	if !ok {
+		return errors.Errorf("address %s is not registered with the address manager", address.TCPAddress())
+	}
+	entry.connectionFailedCount = 0
+	return am.store.updateNotBanned(key, entry)
```

Comment:
Can you add a comment above updateNotBanned function so it will be clear that you update a non-banned address and you don't update an address not to be banned?